tfjs-backend-webgpu/src/kernels/depthwise_conv2d_webgpu.ts

```diff
@@ -104,32 +105,55 @@ export class DepthwiseConv2DProgram implements WebGPUProgram {
         int d1 = d2 / ${channelMul};
         int q = d2 - d1 * ${channelMul};
 
-        int xRCorner = xRCCorner.x;
-        int xCCorner = xRCCorner.y;
+        int inputRowStart = xRCCorner.x;
+        int inputColStart = xRCCorner.y;
+        int inputRowEnd = inputRowStart + ${
+        this.convInfo.filterHeight} * dilation[0];
+        int inputColEnd = inputColStart + ${
+        this.convInfo.filterWidth} * dilation[1];
 
         // Convolve x(?, ?, d1) with w(:, :, d1, q) to get y(yR, yC, d2).
         // ? = to be determined. : = across all values in that axis.
         float dotProd = 0.0;
-        // TODO(xing.xu): Flatten the two for loops and vec4 the operations.
-        for (int wR = 0; wR < filterDims[0]; wR++) {
-          int xR = xRCorner + wR * dilation[0];
 
-          if (xR < 0 || xR >= inDims[0]) {
-            continue;
-          }
+        // Extract if checking out of for loop for performance.
```

Comment:
Here I refered to the tensorflow cuda implementation https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/kernels/depthwise_conv_op_gpu.h#L127